fedimint-server/src/net/connect.rs

```diff
@@ -261,13 +262,167 @@ pub mod mock {
+            }
+            if self.failure_rate.random_fail() {
+                tracing::debug!("Returning random error on unreliable stream");
+                return std::task::Poll::Ready(Err(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    "Randomly failed",
+                )));
+            }
+            std::task::Poll::Ready(Ok(()))
+        }
+    }
+
+    impl AsyncRead for UnreliableDuplexStream {
+        fn poll_read(
+            mut self: Pin<&mut Self>,
+            cx: &mut std::task::Context<'_>,
+            buf: &mut tokio::io::ReadBuf<'_>,
+        ) -> std::task::Poll<std::io::Result<()>> {
+            match self.read_generator.as_mut().map(|g| g.generate(cx)) {
+                Some(std::task::Poll::Ready(Err(e))) => std::task::Poll::Ready(Err(e)),
+                Some(std::task::Poll::Pending) => std::task::Poll::Pending,
+                Some(std::task::Poll::Ready(Ok(()))) | None => {
+                    Pin::new(&mut self.inner).poll_read(cx, buf)
+                }
+            }
+        }
+    }
+
+    impl AsyncWrite for UnreliableDuplexStream {
+        fn poll_write(
+            mut self: Pin<&mut Self>,
+            cx: &mut std::task::Context<'_>,
+            buf: &[u8],
+        ) -> std::task::Poll<Result<usize, std::io::Error>> {
+            match self.write_generator.as_mut().map(|g| g.generate(cx)) {
+                Some(std::task::Poll::Ready(Err(e))) => std::task::Poll::Ready(Err(e)),
+                Some(std::task::Poll::Pending) => std::task::Poll::Pending,
+                Some(std::task::Poll::Ready(Ok(()))) | None => {
+                    Pin::new(&mut self.inner).poll_write(cx, buf)
+                }
+            }
+        }
+
+        fn poll_flush(
+            mut self: Pin<&mut Self>,
+            cx: &mut std::task::Context<'_>,
+        ) -> std::task::Poll<Result<(), std::io::Error>> {
+            match self.flush_generator.as_mut().map(|g| g.generate(cx)) {
+                Some(std::task::Poll::Ready(Err(e))) => std::task::Poll::Ready(Err(e)),
+                Some(std::task::Poll::Pending) => std::task::Poll::Pending,
+                Some(std::task::Poll::Ready(Ok(()))) | None => {
+                    Pin::new(&mut self.inner).poll_flush(cx)
+                }
+            }
+        }
+
+        fn poll_shutdown(
+            mut self: Pin<&mut Self>,
+            cx: &mut std::task::Context<'_>,
+        ) -> std::task::Poll<Result<(), std::io::Error>> {
+            match self.shutdown_generator.as_mut().map(|g| g.generate(cx)) {
+                Some(std::task::Poll::Ready(Err(e))) => std::task::Poll::Ready(Err(e)),
+                Some(std::task::Poll::Pending) => std::task::Poll::Pending,
+                Some(std::task::Poll::Ready(Ok(()))) | None => {
+                    Pin::new(&mut self.inner).poll_shutdown(cx)
+                }
+            }
+        }
+    }
```

Comment:
Wondering if there's a way to not build these futures manually. I'm ok with it in tests, but even a false positive test failure because of a broken futures impl could waste a few hours tracking down.